enhancements/network/ocp-on-smart-nics.md

```diff
@@ -0,0 +1,313 @@
+
+    - Although this enhancement technically only covers the
+      installing/maintaining part of the problem, there will be heavy
+      overlap with the "Smart NIC OVN Offload" enhancement (with the
+      decisions that are made here affecting how OVN and
+      ovn-kubernetes get deployed there).
+
+    - Although not part of the initial design, we assume some
+      customers may be interested taking advantage of the NIC/host
+      separation to run additional security and monitoring
+      functionality on the NICs in the future. Thus, we are interested
+      in designs with more isolation between the NIC and the host,
+      even though that is not necessarily interesting for the primary
+      network-offload use case.
+
+- Design an appropriate architecture for an environment containing a
+  mix of x86_64 hosts with "dumb" NICs, x86_64 hosts with Smart NICs,
+  and the ARM-based Smart NICs themselves. (As discussed below, such
+  an environment may involve one or more OCP clusters.)
+
+- Define how OCP installs and upgrades will work in such an
+  environment.
+
+[Smart NIC OVN Offload]: https://github.com/openshift/enhancements/pull/732
+
+### Non-Goals
+
+- Support for network plugins other than OVN Kubernetes. The
+  architecture designed here may not be generically applicable and is
+  not guaranteed to be usable by other network plugins even if they
+  independently implement their own BlueField, etc, support.
+
+### Supported Hardware
+
+We want to support Smart NICs that are vaguely "BlueField-2-ish". eg:
+
+  1. The NIC has a sufficiently-powerful ARM processor (or,
+     theoretically, a processor of some other architecture that RHCOS
+     supports), and sufficient RAM, SSD, etc, to run as an OCP node.
+
+       - Current BlueField-2 specs mention "Up to 8 Armv8 A72 cores"
+         and "8GB / 16GB / 32GB of on-board DDR4". It is possible we
+         would only support a subset of SKUs.
+
+       - The terms "Smart NIC" and "DPU" are also being used to refer
+         to NICs that just have programmable ASICs or FPGAs, rather
+         than a full computing environment. These are out of scope.
+
+  2. The NIC has an IPMI-addressable Baseboard Management Controller
+     that can be used to control and provision it independently of
+     its host.
+
+  3. The NIC implements SR-IOV in such a way that we can connect
+     pods on the host directly to an OVN bridge on the NIC.
+
+  4. All of the above can be done without running any proprietary
+     drivers or third-party software on either the host's or the NIC's
+     RHCOS installation.
+
+## Design Details
+
+### Clusters Will Contain a Mix of Smart and "Dumb" NICs
+
+OVN Kubernetes is not currently capable of mixing offloaded and
+non-offloaded networking on the same node, so every pod on a node
+containing a Smart NIC must have offloaded networking, which, in the
+current architecture, means it must explicitly _request_ offloaded
```

Comment:
@danwinship I think we should try to change this. It's not clear to me (though I haven't read deeply) why a pod would care that it was offloaded or not. So sure, pods could request being scheduled to an offloaded node, but by default I'd presume they'd land somewhere and if that happened to be a smartnic node, so be it.

Reply:
I talked about this a little with @zshi-redhat ; it seems that the problem is that we need to use the sr-iov device plugin to allocate a VF, but kubelet will only ask it to do that if the pod has the right annotations, and we can't just annotate it via an admission plugin, because the admission plugin runs before the pod gets scheduled. But I think he agreed it would be possible to change things to have the offloaded-ovn-kube cni plugin do the VF allocation itself

Reply:
Yes, pod can request being scheduled to an offloaded node by adding sriov extended resource request in pod spec, on the other hand, pod that doesn't request extended resource can also be scheduled to the offloaded node unless we explicitly set affinity or labels or taints. 
The problem now is there are several cluster pods or infra pods such as pods in openshift-monitoring, openshift-ingress namespaces etc which don't have affinity or labels or taints that allow them to not be scheduled on the offloaded node. So if the scheduling of these pods to offloaded node happens, then ovn-k8s is not capable of creating network for these pods which results in failure of pod creation.
There are several ways to solve this:
1) run these infra pods with hostnetwork:true  which doesn't need to run ovn-k8s CNI
2) add extended sriov resource request to these pods that *will* be scheduled on the offloaded node (which may not be able to know upfront before k8s-scheduler runs)
3) set anti-affinity to the daemonsets/deployments so that their pods won't be scheduled on offloaded node.
All above three would require changing the manifests in bindata for various openshift cluster operators.

Reply:
@zshi-redhat can you please elaborate what's the difference between openshift-monitoring, ingress and others? What they expected to have? Can you please provide an example. If it's a minor change, I guess we can ask relevant teams to push it forward.

Reply:
@romfreiman With current OVN-Kubernetes design, it is only supported to run SR-IOV pods on the SmartNIC x86 node. Regular pods whose networks require virtual **veth** interfaces cannot be plugged to OVN network because there will be no Open vSwitch running on the SmartNIC x86 node. Some pods in openshift-monitoring, openshift-ingress or other namespaces are the pods whose networks require virtual **veth** interfaces. So this is a limitation that current SmartNIC x86 nodes cannot host regular **veth** pods.

The example of setting pod to use host network is as below:
```
apiVersion: v1
kind: Pod
metadata:
  name: testpod
spec:
  hostNetwork: true       <======
  containers:
  - name: container0
    image: centos
    command: [ "/bin/bash", "-c", "--" ]
    args: [ "while true; do sleep 300000; done;" ]
```
It needs change in pod or daemonset or deployment manifests.
This problem is also described [here](https://github.com/openshift/enhancements/pull/732/files#diff-9cb730b0bcd44036b71650e25bd939ec20d549835f76ff5b1e964c6d20d6e574R277-R340).